extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -536,37 +574,152 @@ export class AmpStoryEmbeddedComponent {
+   * @private
+   */
+  closeExpandedEl_() {
+    this.triggeringTarget_.classList.toggle(
+        'i-amphtml-expanded-component', false);
+    const embedId =
+      this.triggeringTarget_.getAttribute(EMBED_ID_ATTRIBUTE_NAME);
+
+    const styleEl = embedStyleEls[embedId];
+
+    const transformString = styleEl.textContent.substring(
+        styleEl.textContent.indexOf('transform'),
+        styleEl.textContent.indexOf('margin'));
+
+    embedStyleEls[embedId].textContent =
+      styleEl.textContent.replace(transformString,
+          `transform: scale(${styleEl[AMP_EMBED_DATA].scale}) !important;`);
+  }
+
+  /**
+   * Animates into expanded view. It calculates what the full-screen dimensions
+   * of the element will be, and uses them to deduce the translateX/Y values
+   * once the element reaches its full-screen size.
    * @param {!Element} target
    * @private
    */
   animateExpanded_(target) {
+    const embedId = target.getAttribute(EMBED_ID_ATTRIBUTE_NAME);
     const state = {};
+    const storedStyle = embedStyleEls[embedId][AMP_EMBED_DATA];
     this.resources_.measureMutateElement(target,
         /** measure */
         () => {
           const targetRect = target./*OK*/getBoundingClientRect();
+          // TODO(#20832): Store DOMRect for the page in the store to avoid
+          // having to call getBoundingClientRect().
           const pageRect = this.componentPage_./*OK*/getBoundingClientRect();
 
-          const centeredTop = pageRect.height / 2 - targetRect.height / 2;
-          const centeredLeft = pageRect.width / 2 - targetRect.width / 2;
+          // Gap on the left of the element between full-screen size and
+          // current size.
+          const leftGap = (storedStyle.width - targetRect.width) / 2;
+          // Distance from left of page to what will be the left of the
+          // element in full-screen.
+          const fullScreenLeft = targetRect.left - leftGap - pageRect.left;
+          const centeredLeft = pageRect.width / 2 - storedStyle.width / 2;
+          state.translateX = centeredLeft - fullScreenLeft;
+
+          // Gap on the top of the element between full-screen size and
+          // current size.
+          const topGap = (storedStyle.height - targetRect.height) / 2;
+          // Distance from top of page to what will be the top of the element in
+          // full-screen.
+          const fullScreenTop = targetRect.top - topGap - pageRect.top;
+          const centeredTop = pageRect.height / 2 - storedStyle.height / 2;
+          state.translateY = centeredTop - fullScreenTop;
+        },
+        /** mutate */
+        () => {
+          target.classList.toggle('i-amphtml-expanded-component', true);
+
+          const styleEl = embedStyleEls[embedId];
+          embedStyleEls[embedId].textContent = styleEl.textContent.split(
+              `scale(${storedStyle.scale})`).join(`scale(1)
+              translate3d(${state.translateX}px, ${state.translateY}px, 0)`);
+        });
```

Comment:
It's too late now, but I'm getting very confused by the strings manipulation to figure out the styles. They're spread out throughout the file and they all do different things. Figuring out which part of the file sets which CSS property is very hard.

One idea that would've made things easier:

1. Each `<style>` has its `embedData`
2. One method accepts `embedData` as a parameter, and updates the textContent of `<style>`
3. Now the logic is centralized: to update the styles, get the `embedData`, change what you want updated, and call the one method

Maybe in a future refactoring :)

Reply:
I had actually suggested something similar offline; I think it shouldn't be too bad to address this in this PR.

Reply:
Done.